# Segmenting and Clustering Neighborhoods in Toronto - Part III

<hr style="height:5px;background-color:black">

## 1. Data preparation

### Let's import some libraries for the purpose of this exercise

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs:
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    geopy-1.21.0               |             py_0          58 KB  conda-forge
    openssl-1.1.1g             |       h516909a_0         2.1 MB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.2 MB

The following NEW packages will be INSTALLED:

  geographiclib      conda-forge/noarch::geographiclib-1.50-py_0
  geopy              conda-forge/noarch::geopy-1.21.0-py_0

The following packages will be UPDATED:

  openssl                                 1.1.1f-h516909a_0 --> 1.1.1g-h516909a_0



openssl-1.1.1g       | 2.1 MB    | ############

### Let's download the dataframe we built in Part II

I uploaded it on my GitHub repository dedicated to this assignment:
https://raw.githubusercontent.com/oliviercy/Capstone_Labs/master/Part2_neighborhoods_coo_dataframe.csv

In [2]:
!wget -O Part2_neighborhoods_coo_dataframe.csv https://raw.githubusercontent.com/oliviercy/Capstone_Labs/master/Part2_neighborhoods_coo_dataframe.csv

In [3]:
torontoz = pd.read_csv('Part2_neighborhoods_coo_dataframe.csv')
torontoz.head()

PostalCode           Borough                                 Neighborhood  \
0        M3A        North York                                    Parkwoods   
1        M4A        North York                             Victoria Village   
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
3        M6A        North York             Lawrence Manor, Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [4]:
torontoz.shape

(103, 5)

In [5]:
# Tip 2
torontoz['Borough'].nunique()

10

#### There are <span style="color:Crimson">103</span> Postal codes in the <span style="color:Crimson">10</span> Boroughs of Toronto

### Let's use geopy library to get the latitude and longitude values of Toronto

In order to define an instance of the geocoder, we need to define a user_agent named <em>toronto_explorer</em>, as shown below.

In [6]:
address = 'Toronto, ON, Canada'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geographical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geographical coordinate of Toronto are 43.6534817, -79.3839347.


### Let' create a map of Toronto with neighborhoods superimposed on top

In [7]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(torontoz['Latitude'], torontoz['Longitude'], torontoz['Borough'], torontoz['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

<hr style="height:5px;background-color:Crimson">

## 2. Foursquare API - Explore venues

### Let's define Foursquare credentials and version

In [8]:
CLIENT_ID = 'OFUHCKAEJDYXI2WQ2HTRI0MYSSNYUHRYFEEAAWMGA4YQ1LHU' # Foursquare ID
CLIENT_SECRET = 'PS2SCDFEPJIBXHFYO0CDCTHXCDHQKD1EKBZONRZN1TEPCJR1' # Foursquare Secret
VERSION = '20180605' # Foursquare API version

### Let's create a function to get the top 100 venues that are within a radius of 500 meters of each neighborhood

In [9]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    LIMIT = 100
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius,
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

### This cell will run the above function on each neighborhood and create a new dataframe called `toronto_venues`

In [10]:
toronto_venues = getNearbyVenues(names=torontoz['Neighborhood'],
                                   latitudes=torontoz['Latitude'],
                                   longitudes=torontoz['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview
The Danforth West, Ri

In [11]:
print(toronto_venues.shape)
toronto_venues.head()

(2151, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2  Victoria Village              43.725882              -79.315572   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0         Brookbanks Park       43.751976       -79.332140   
1           Variety Store       43.751974       -79.333114   
2  Victoria Village Arena       43.723481       -79.315635   
3             Tim Hortons       43.725517       -79.313103   
4               Portugril       43.725819       -79.312785   

          Venue Category  
0                   Park  
1      Food & Drink Shop  
2           Hockey Arena  
3            Coffee Shop  
4  Portuguese Restaurant

### Let's check how many venues were returned for each neighborhood

In [12]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               5   
Alderwood, Long Branch                                                 10   
Bathurst Manor, Wilson Heights, Downsview North                        19   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      24   
Berczy Park                                                            57   
Birch Cliff, Cliffside West                                             4   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre                                  18   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Caledonia-Fairbanks                                                     4   
Canada Post Gateway Processing Centre                                  12   
Cedarbrae                                                               8   
Central Bay Street                                                     63   
Christie                                                               17   
Church and Wellesley                                                   78   
Clarks Corners, Tam O'Shanter, Sullivan                                15   
Cliffside, Cliffcrest, Scarborough Village West                         2   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             36   
Davisville North                                                        8   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                        4   
Don Mills                                                              27   
Dorset Park, Wexford Heights, Scarborough Town ...                      7   
Downsview                                                              15   
Dufferin, Dovercourt Village                                           19   
East Toronto                                                            3   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                      9   
Fairview, Henry Farm, Oriole                                           70   
First Canadian Place, Underground city                                100   
Forest Hill North & West                                                5   
Garden District, Ryerson                                              100   
Glencairn                                                               5   
Golden Mile, Clairlea, Oakridge                                        10   
Guildwood, Morningside, West Hill                                       7   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          23   
Hillcrest Village                                                       5   
Humber Summit                                                           1   
Humberlea, Emery                                                        2   
Humewood-Cedarvale                                                      4   
India Bazaar, The Beaches West                                         20   
Kennedy Park, Ionview, East Birchmount Park                             6   
Kensington Market, Chinatown, Grange Park                              55   
Kingsview Village, St. Phillips, Martin Grove G...                      4   
Lawrence Manor, Lawrence Heights                                       17   
Lawrence Park                                                           4   
Leaside                                                                32   
Little Portugal, Trinity                                               41   
Malvern, Rouge                                                          2   
Milliken, Agincourt North, Steeles East, L'Amor.

### Let's find out how many unique categories can be curated from all the returned venues

In [13]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 267 uniques categories.


## 3. Analyze each Neighborhood

In [14]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# Move neighborhood column to the first column - Tip 1
## get a list of columns
cols = list(toronto_onehot)

## move the column to head of list using index, pop and insert
cols.insert(0, cols.pop(cols.index('Neighborhood')))

## use loc to reorder
toronto_onehot = toronto_onehot.loc[:, cols]
toronto_onehot.head()

toronto_onehot.head()

Neighborhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2  Victoria Village                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
0                    0             0         0            0           0   
1                    0             0         0            0           0   
2                    0             0         0            0           0   
3                    0             0         0            0           0   
4                    0             0         0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Garage  \
0                    0                 0                   0            0   
1                    0                 0                   0            0   
2                    0                 0                   0            0   
3                    0                 0                   0            0   
4                    0                 0                   0            0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Field  Baseball Stadium  Basketball Court  Basketball Stadium  \
0               0                 0                 0                   0   
1               0                 0                 0                   0   
2               0                 0                 0                   0   
3               0                 0                 0                   0   
4               0                 0                 0                   0   

   Beach  Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  \
0      0                0         0           0                   0   
1      0                0         0           0                   0   
2      0                0         0           0                   0   
3      0                0         0           0                   0   
4      0                0         0           0                   0   

   Bike Shop  Bistro  Boat or Ferry  Bookstore  Boutique  \
0          0       0              0          0         0   
1          0       0              0          0         0   
2          0       0              0          0         0   
3          0       0              0          0         0   
4          0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0        0            0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0                0         0             0              0         0   
1               

And let's examine the new dataframe size.

In [15]:
toronto_onehot.shape

(2151, 267)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [16]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0               0.0             0.000000           0.0       0.0          0.0   
1               0.0             0.000000           0.0       0.0          0.0   
2               0.0             0.000000           0.0       0.0          0.0   
3               0.0             0.000000           0.0       0.0          0.0   
4               0.0             0.041667           0.0       0.0          0.0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0         0.0                  0.0               0.0                 0.0   
1         0.0                  0.0               0.0                 0.1   
2         0.0                  0.0               0.0                 0.0   
3         0.0                  0.0               0.0                 0.0   
4         0.0                  0.0               0.0                 0.0   

   Auto Garage  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  \
0          0.0            0.0        0.0         0.0         0.0     0.0   
1          0.0            0.0        0.0         0.0         0.0     0.0   
2          0.0            0.0        0.0         0.0         0.0     0.0   
3          0.0            0.0        0.0         0.0         0.0     0.0   
4          0.0            0.0        0.0         0.0         0.0     0.0   

       Bank  Bar  Baseball Field  Baseball Stadium  Basketball Court  \
0  0.000000  0.0             0.0               0.0               0.0   
1  0.000000  0.0             0.0               0.0               0.0   
2  0.105263  0.0             0.0               0.0               0.0   
3  0.250000  0.0             0.0               0.0               0.0   
4  0.000000  0.0             0.0               0.0               0.0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                 0.0    0.0              0.0       0.0         0.0   
1                 0.0    0.0              0.0       0.0         0.0   
2                 0.0    0.0              0.0       0.0         0.0   
3                 0.0    0.0              0.0       0.0         0.0   
4                 0.0    0.0              0.0       0.0         0.0   

   Belgian Restaurant  Bike Shop  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                 0.0        0.0     0.0            0.0        0.0       0.0   
1                 0.0        0.0     0.0            0.0        0.0       0.0   
2                 0.0        0.0     0.0            0.0        0.0       0.0   
3                 0.0        0.0     0.0            0.0        0.0       0.0   
4                 0.0        0.0     0.0            0.0        0.0       0.0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                   0.0             0.2      0.0     0.000000   
1                   0.0             0.0      0.0     0.000000   
2                   0.0             0.0      0.0     0.052632   
3                   0.0             0.0      0.0     0.000000   
4                   0.0             0.0      

#### Let's confirm the new size

In [17]:
toronto_grouped.shape

(93, 267)

#### Let's print each neighborhood along with the top 5 most common venues

In [18]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                       venue  freq
0  Latin American Restaurant   0.2
1             Breakfast Spot   0.2
2                     Lounge   0.2
3               Skating Rink   0.2
4             Clothing Store   0.2


----Alderwood, Long Branch----
                venue  freq
0         Pizza Place   0.2
1        Dance Studio   0.1
2        Skating Rink   0.1
3      Sandwich Place   0.1
4  Athletics & Sports   0.1


----Bathurst Manor, Wilson Heights, Downsview North----
           venue  freq
0           Bank  0.11
1    Coffee Shop  0.11
2       Pharmacy  0.05
3          Diner  0.05
4  Shopping Mall  0.05


----Bayview Village----
                 venue  freq
0  Japanese Restaurant  0.25
1                 Café  0.25
2                 Bank  0.25
3   Chinese Restaurant  0.25
4               Museum  0.00


----Bedford Park, Lawrence Manor East----
                venue  freq
0          Restaurant  0.08
1    Sushi Restaurant  0.08
2  Italian Restaurant  0.08
3         Coffee Sh

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [19]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [20]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East        Sandwich Place   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0          Skating Rink  Latin American Restaurant        Breakfast Spot   
1        Sandwich Place               Skating Rink              Pharmacy   
2                  Bank        Fried Chicken Joint           Bridal Shop   
3                  Bank         Chinese Restaurant   Japanese Restaurant   
4    Italian Restaurant                 Restaurant           Coffee Shop   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Clothing Store           Yoga Studio             Drugstore   
1                   Gym          Dance Studio           Coffee Shop   
2        Sandwich Place                 Diner            Restaurant   
3           Yoga Studio        Discount Store   Distribution Center   
4      Sushi Restaurant      Greek Restaurant       Thai Restaurant   

       8th Most Common Venue    9th Most Common Venue 10th Most Common Venue  
0        Distribution Center                  Dog Run       Doner Restaurant  
1         Athletics & Sports                      Pub     Dim Sum Restaurant  
2  Middle Eastern Restaurant              Supermarket       Sushi Restaurant  
3                    Dog Run         Doner Restaurant             Donut Shop  
4               Liquor Store  Comfort Food Restaurant              Juice Bar

## 4. Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [23]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:93]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1,
       1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 2, 1, 3, 1, 1, 1, 1, 1, 4, 1, 1, 3, 1, 1, 1, 3,
       1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3,
       1, 1, 1, 1, 3], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [52]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = torontoz

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head(6) # check the last columns!

PostalCode           Borough                                 Neighborhood  \
0        M3A        North York                                    Parkwoods   
1        M4A        North York                             Victoria Village   
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
3        M6A        North York             Lawrence Manor, Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
5        M9A         Etobicoke                             Islington Avenue   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             3.0     Food & Drink Shop   
1  43.725882 -79.315572             1.0           Pizza Place   
2  43.654260 -79.360636             1.0           Coffee Shop   
3  43.718518 -79.464763             1.0        Clothing Store   
4  43.662301 -79.389494             1.0           Coffee Shop   
5  43.667856 -79.532242             NaN                   NaN   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                    Park             Drugstore                 Diner   
1             Coffee Shop          Intersection     French Restaurant   
2                  Bakery                   Pub                  Park   
3  Furniture / Home Store           Coffee Shop           Event Space   
4        Sushi Restaurant                 Diner          Burger Joint   
5                     NaN                   NaN                   NaN   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0         Discount Store   Distribution Center                      Dog Run   
1  Portuguese Restaurant          Hockey Arena  Eastern European Restaurant   
2         Breakfast Spot                  Café                      Theater   
3             Shoe Store   Sporting Goods Shop           Miscellaneous Shop   
4          Burrito Place             Juice Bar                         Café   
5                    NaN                   NaN                          NaN   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant            Donut Shop            Yoga Studio  
1     Electronics Store             Drugstore           Dessert Shop  
2           Yoga Studio        Cosmetics Shop             Shoe Store  
3   Arts & Crafts Store     Accessories Store               Boutique  
4   Japanese Restaurant    Italian Restaurant               Beer Bar  
5                   NaN                   NaN                    NaN

In [53]:
toronto_merged.shape

(103, 16)

In [54]:
# Tip 3
# Let's check for NaN values
df_nan = toronto_merged[toronto_merged.isna().any(axis=1)]
df_nan

PostalCode      Borough                                       Neighborhood  \
5         M9A    Etobicoke                                   Islington Avenue   
11        M9B    Etobicoke  West Deane Park, Princess Gardens, Martin Grov...   
45        M2L   North York                           York Mills, Silver Hills   
52        M2M   North York                            Willowdale, Newtonbrook   
95        M1X  Scarborough                                        Upper Rouge   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
5   43.667856 -79.532242             NaN                   NaN   
11  43.650943 -79.554724             NaN                   NaN   
45  43.757490 -79.374714             NaN                   NaN   
52  43.789053 -79.408493             NaN                   NaN   
95  43.836125 -79.205636             NaN                   NaN   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
5                    NaN                   NaN                   NaN   
11                   NaN                   NaN                   NaN   
45                   NaN                   NaN                   NaN   
52                   NaN                   NaN                   NaN   
95                   NaN                   NaN                   NaN   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
5                    NaN                   NaN                   NaN   
11                   NaN                   NaN                   NaN   
45                   NaN                   NaN                   NaN   
52                   NaN                   NaN                   NaN   
95                   NaN                   NaN                   NaN   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
5                    NaN                   NaN                    NaN  
11                   NaN                   NaN                    NaN  
45                   NaN                   NaN                    NaN  
52                   NaN                   NaN                    NaN  
95                   NaN                   NaN                    NaN

In [55]:
# Let's drop the lines with NaN values
toronto_merged = toronto_merged.dropna()
toronto_merged.shape

(98, 16)

Finally, let's visualize the resulting clusters

In [58]:
# Tip 4
# First, let's cast the 'Cluster Labels' column from float to integer
toronto_merged = toronto_merged.astype({'Cluster Labels': 'int32'})

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=10)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

<a id='item5'></a>

## 5. Examine Clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.

#### Cluster 1

In [59]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
51  Scarborough               0   American Restaurant                 Motel   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
51             Drugstore                 Diner        Discount Store   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
51   Distribution Center               Dog Run      Doner Restaurant   

   9th Most Common Venue 10th Most Common Venue  
51            Donut Shop            Yoga Studio

#### Cluster 2

In [60]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
1          North York               1                 Pizza Place   
2    Downtown Toronto               1                 Coffee Shop   
3          North York               1              Clothing Store   
4    Downtown Toronto               1                 Coffee Shop   
6         Scarborough               1        Fast Food Restaurant   
7          North York               1                         Gym   
8           East York               1                 Pizza Place   
9    Downtown Toronto               1              Clothing Store   
10         North York               1                 Pizza Place   
12        Scarborough               1  Construction & Landscaping   
13         North York               1                         Gym   
14          East York               1                        Park   
15   Downtown Toronto               1                        Café   
16               York               1                       Field   
17          Etobicoke               1                 Pizza Place   
18        Scarborough               1          Mexican Restaurant   
19       East Toronto               1           Health Food Store   
20   Downtown Toronto               1                 Coffee Shop   
22        Scarborough               1                 Coffee Shop   
23          East York               1         Sporting Goods Shop   
24   Downtown Toronto               1                 Coffee Shop   
25   Downtown Toronto               1               Grocery Store   
26        Scarborough               1        Caribbean Restaurant   
27         North York               1          Athletics & Sports   
28         North York               1                 Coffee Shop   
29          East York               1           Indian Restaurant   
30   Downtown Toronto               1                 Coffee Shop   
31       West Toronto               1                    Pharmacy   
33         North York               1              Clothing Store   
34         North York               1              Massage Studio   
36   Downtown Toronto               1                 Coffee Shop   
37       West Toronto               1                         Bar   
38        Scarborough               1              Discount Store   
39         North York               1                        Café   
40         North York               1               Grocery Store   
41       East Toronto               1            Greek Restaurant   
42   Downtown Toronto               1                 Coffee Shop   
43       West Toronto               1                        Café   
44        Scarborough               1                    Bus Line   
46         North York               1               Grocery Store   
47       East Toronto               1        Fast Food Restaurant   
48   Downtown Toronto               1                 Coffee Shop   
49         North York               1            Basketball Court   
50         North York               1                 Pizza Place   
53         North York               1               Grocery Store   
54       East Toronto               1                        Café   
55         North York               1              Sandwich Place   
56               York               1        Fast Food Restaurant   
58        Scarborough               1             College Stadium   
59         North York               1                 Coffee Shop   
60         North York               1               Grocery Store   
61    Central Toronto               1                        Park   
62    Central Toronto               1                      Garden   
63               York               1        Caribbean Restaurant   
65        Scarborough               1           Indian Restaurant   
67    Central Toronto               1                         Gym   
68    Central Toronto               1               Jewelry Store   
69       Wes

#### Cluster 3

In [61]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
32  Scarborough               2            Playground           Yoga Studio   
35    East York               2                  Park           Coffee Shop   
85  Scarborough               2            Playground                  Park   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
32            Donut Shop    Dim Sum Restaurant                 Diner   
35     Convenience Store           Yoga Studio            Donut Shop   
85           Coffee Shop           Yoga Studio            Donut Shop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
32        Discount Store   Distribution Center               Dog Run   
35                 Diner        Discount Store   Distribution Center   
85                 Diner        Discount Store   Distribution Center   

   9th Most Common Venue 10th Most Common Venue  
32      Doner Restaurant              Drugstore  
35               Dog Run       Doner Restaurant  
85               Dog Run       Doner Restaurant

#### Cluster 4

In [62]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York               3     Food & Drink Shop   
21              York               3                  Park   
64              York               3                  Park   
66        North York               3                  Park   
83   Central Toronto               3                  Park   
91  Downtown Toronto               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                   Park             Drugstore                 Diner   
21         Women's Store                   Spa             Drugstore   
64           Yoga Studio             Drugstore                 Diner   
66     Convenience Store                  Bank                   Bar   
83           Yoga Studio             Drugstore                 Diner   
91            Playground                 Trail           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0         Discount Store   Distribution Center               Dog Run   
21        Discount Store   Distribution Center               Dog Run   
64        Discount Store   Distribution Center               Dog Run   
66           Yoga Studio             Drugstore   Distribution Center   
83        Discount Store   Distribution Center               Dog Run   
91      Doner Restaurant    Dim Sum Restaurant                 Diner   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0       Doner Restaurant            Donut Shop                  Yoga Studio  
21      Doner Restaurant            Donut Shop            Electronics Store  
64      Doner Restaurant            Donut Shop  Eastern European Restaurant  
66               Dog Run      Doner Restaurant                   Donut Shop  
83      Doner Restaurant            Donut Shop  Eastern European Restaurant  
91        Discount Store   Distribution Center                      Dog Run

#### Cluster 5

In [63]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels          1st Most Common Venue  \
57   North York               4  Paper / Office Supplies Store   
101   Etobicoke               4                 Baseball Field   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
57         Baseball Field           Yoga Studio             Drugstore   
101           Yoga Studio             Drugstore        Discount Store   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
57         Discount Store   Distribution Center               Dog Run   
101   Distribution Center               Dog Run      Doner Restaurant   

    8th Most Common Venue        9th Most Common Venue  \
57       Doner Restaurant                   Donut Shop   
101            Donut Shop  Eastern European Restaurant   

          10th Most Common Venue  
57   Eastern European Restaurant  
101          Filipino Restaurant

<hr style="height:3px;background-color:black">

# Acknowledgments: list of tips

1. Move column by name: <br />
https://stackoverflow.com/questions/25122099/move-column-by-name-to-front-of-table-in-pandas

2. Count unique values per group: <br />
https://stackoverflow.com/questions/38309729/count-unique-values-with-pandas-per-groups/38309823

3. Display rows with NaN values: <br />
https://stackoverflow.com/questions/43424199/display-rows-with-one-or-more-nan-values-in-pandas-dataframe

4. Change the data type of a column: <br />
https://stackoverflow.com/questions/15891038/change-data-type-of-columns-in-pandas